src/sitemap-pipe.js

```diff
@@ -20,6 +21,48 @@ import setCustomResponseHeaders from './steps/set-custom-response-headers.js';
 import { PipelineStatusError } from './PipelineStatusError.js';
 import { PipelineResponse } from './PipelineResponse.js';
 
+async function generateSitemap(state) {
+  const {
+    owner, repo, ref, contentBusId, partition, s3Loader, log,
+    previewHost, config: { host: prodCDN } = {},
+  } = state;
+  const ret = await s3Loader.getObject('helix-content-bus', `${contentBusId}/live/sitemap.json`);
+  if (ret.status !== 200) {
+    return ret;
+  }
+  let config;
+  try {
+    config = JSON.parse(ret.body);
+  } catch (e) {
+    log.info('failed to parse /sitemap.json', e);
+    throw new PipelineStatusError(404, `Failed to parse /sitemap.json: ${e.message}`);
+  }
+  const { data } = config;
+  if (!data || !Array.isArray(data)) {
+    throw new PipelineStatusError(404, 'Expected \'data\' array not found in /sitemap.json');
+  }
+  const host = partition === 'preview'
+    ? (previewHost || `${ref}--${repo}--${owner}.hlx.page`)
+    : (prodCDN || `${ref}--${repo}--${owner}.hlx.live`);
+  const loc = ({ path, lastModified }) => `  <url>
+    <loc>
+      https://${host}${escape(path)}
+    </loc>
+    <lastmod>${new Date(lastModified * 1000).toISOString().substring(0, 10)}</lastmod>
+  </url>`;
+  const xml = `<?xml version="1.0" encoding="utf-8"?>
+<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:xhtml="http://www.w3.org/1999/xhtml">
+${data.map((record) => loc(record)).join('\n')}
+</urlset>
+`;
```

Comment:
isn't it faster (and nicer to read) to create an array of all lines and then join at the end?

```
[
  '<?xml version="1.0" encoding="utf-8"?>',
  '<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:xhtml="http://www.w3.org/1999/xhtml">',
  ...data..map((record) => loc(record)),
  '</urlset>'
].join('\n);
```